# 知能システム理論基礎　レポート課題3　ナイーブベイズ

筑波大学

知能機能システム　202120812 HAO YIYANG

  

---






**1.    自分で２種類のテキストデータを準備し、分類しなさい**





まず、課題の準備として、scikit-learn、Janome、scipy の三つのライブラリをインストールして用意する。

In [ ]:
!pip install scikit-learn
!pip install janome
!pip install scipy

     |████████████████████████████████| 19.7 MB 1.3 MB/s 


ここで、自分で準備したテキストデータを訓練用テキストデータとして使用する。

私は、青空文庫から太宰治の「人間失格」とWikipediaで用語「単純ベイズ分類器」の解説を利用して、ナイーブベイズの実験を行う。



*   https://www.aozora.gr.jp/cards/000035/files/301_14912.html
*   https://ja.wikipedia.org/wiki/%E5%8D%98%E7%B4%94%E3%83%99%E3%82%A4%E3%82%BA%E5%88%86%E9%A1%9E%E5%99%A8



In [ ]:
from google.colab import files
files.upload()

Saving ngsk.txt to ngsk.txt
Saving wiki.txt to wiki.txt


{'ngsk.txt': b'\xe7\xa7\x81\xe3\x81\xaf\xe3\x80\x81\xe3\x81\x9d\xe3\x81\xae\xe7\x94\xb7\xe3\x81\xae\xe5\x86\x99\xe7\x9c\x9f\xe3\x82\x92\xe4\xb8\x89\xe8\x91\x89\xe3\x80\x81\xe8\xa6\x8b\xe3\x81\x9f\xe3\x81\x93\xe3\x81\xa8\xe3\x81\x8c\xe3\x81\x82\xe3\x82\x8b\xe3\x80\x82\r\n\xe3\x80\x80\xe4\xb8\x80\xe8\x91\x89\xe3\x81\xaf\xe3\x80\x81\xe3\x81\x9d\xe3\x81\xae\xe7\x94\xb7\xe3\x81\xae\xe3\x80\x81\xe5\xb9\xbc\xe5\xb9\xb4\xe6\x99\x82\xe4\xbb\xa3\xe3\x80\x81\xe3\x81\xa8\xe3\x81\xa7\xe3\x82\x82\xe8\xa8\x80\xe3\x81\x86\xe3\x81\xb9\xe3\x81\x8d\xe3\x81\xa7\xe3\x81\x82\xe3\x82\x8d\xe3\x81\x86\xe3\x81\x8b\xe3\x80\x81\xe5\x8d\x81\xe6\xad\xb3\xe5\x89\x8d\xe5\xbe\x8c\xe3\x81\x8b\xe3\x81\xa8\xe6\x8e\xa8\xe5\xae\x9a\xe3\x81\x95\xe3\x82\x8c\xe3\x82\x8b\xe9\xa0\x83\xe3\x81\xae\xe5\x86\x99\xe7\x9c\x9f\xe3\x81\xa7\xe3\x81\x82\xe3\x81\xa3\xe3\x81\xa6\xe3\x80\x81\xe3\x81\x9d\xe3\x81\xae\xe5\xad\x90\xe4\xbe\x9b\xe3\x81\x8c\xe5\xa4\xa7\xe5\x8b\xa2\xe3\x81\xae\xe5\xa5\xb3\xe3\x81\xae\xe3\x81\xb2\xe3\x81\xa8\xe3\x81\

!lsコマンドを使って、ファイルのアップロード状況を確認する。

In [ ]:
!ls

iit_naive_bayes.py  ngsk.txt  __pycache__  sample_data	wiki.txt


２つ訓練用データは日本語であるから、ここで日本語を指定する。

In [ ]:
%reload_ext autoreload
%autoreload 2
import iit_naive_bayes as iitnb
nb = iitnb.IITNaiveBayses("ngsk.txt", "wiki.txt","jpn")

２つの訓練データを確認する

In [ ]:
nb.print_file1()

私は、その男の写真を三葉、見たことがある。
　一葉は、その男の、幼年時代、とでも言うべきであろうか、十歳前後かと推定される頃の写真であって、その子供が大勢の女のひとに取りかこまれ、
（それは、その子供の姉たち、妹たち、それから、従姉妹いとこたちかと想像される）庭園の池のほとりに、荒い縞の袴はかまをはいて立ち、首を三十度ほど左に傾け、
醜く笑っている写真である。醜く？　けれども、鈍い人たち（つまり、美醜などに関心を持たぬ人たち）は、面白くも何とも無いような顔をして、
「可愛い坊ちゃんですね」
　といい加減なお世辞を言っても、まんざら空からお世辞に聞えないくらいの、謂いわば通俗の「可愛らしさ」みたいな影もその子供の笑顔に無いわけではないのだが、
しかし、いささかでも、美醜に就いての訓練を経て来たひとなら、ひとめ見てすぐ、
「なんて、いやな子供だ」
　と頗すこぶる不快そうに呟つぶやき、毛虫でも払いのける時のような手つきで、その写真をほうり投げるかも知れない。
　まったく、その子供の笑顔は、よく見れば見るほど、何とも知れず、イヤな薄気味悪いものが感ぜられて来る。どだい、それは、笑顔でない。
この子は、少しも笑ってはいないのだ。その証拠には、この子は、両方のこぶしを固く握って立っている。人間は、こぶしを固く握りながら笑えるものでは無いのである。
猿だ。猿の笑顔だ。ただ、顔に醜い皺しわを寄せているだけなのである。「皺くちゃ坊ちゃん」とでも言いたくなるくらいの、まことに奇妙な、そうして、
どこかけがらわしく、へんにひとをムカムカさせる表情の写真であった。私はこれまで、こんな不思議な表情の子供を見た事が、いちども無かった。



In [ ]:
nb.print_file2()

概要
　単純ベイズ分類器の元となる確率モデルは強い（単純な）独立性仮定と共にベイズの定理を適用することに基づいており、
より正確に言えば「独立特徴モデル」と呼ぶべきものである。

　確率モデルの性質に基づいて、単純ベイズ分類器は教師あり学習の設定で効率的に訓練可能である。多くの実用例では、
単純ベイズ分類器のパラメータ推定には最尤法が使われる。つまり、単純ベイズ分類器を使用するにあたって、ベイズ確率やその他のベイズ的手法を使う必要はない。

　設計も仮定も非常に単純であるにも関わらず、単純ベイズ分類器は複雑な実世界の状況において、期待よりもずっとうまく働く。
近頃、ベイズ分類問題の注意深い解析によって、単純ベイズ分類器の効率性に理論的理由があることが示された。
単純ベイズ分類器の利点は、分類に不可欠なパラメータ（変数群の平均と分散）を見積もるのに、訓練例データが少なくて済む点である。
変数群は独立であると仮定されているため、各クラスについての変数の分散だけが必要であり、共分散行列全体は不要である。



そこで、以下の３つテキストを用いて、ナイーブベイズを利用して、テキスト分類を試し見る。



1.   第二葉の写真の顔は、これはまた、びっくりするくらいひどく変貌していた。
2.   抽象的には、分類器の確率モデルは次のような依存クラス変数Cについての条件付きモデルである。
3.   斜め七十七度の並びで泣く泣くいななくナナハン七台難なく並べて長眺め

（1は人間失格の原文、2はWiki用語説明の一部、3は両方とも所属しないテキストである）




In [ ]:
nb.predict_proba("第二葉の写真の顔は、これはまた、びっくりするくらいひどく変貌していた。")

array([[9.99441308e-01, 5.58692399e-04]])

In [ ]:
nb.predict_proba("抽象的には、分類器の確率モデルは次のような依存クラス変数Cについての条件付きモデルである。")

array([[2.09948334e-06, 9.99997901e-01]])

In [ ]:
nb.predict_proba("斜め七十七度の並びで泣く泣くいななくナナハン七台難なく並べて長眺め")

array([[0.88538038, 0.11461962]])

分類の結果として、テストテキスト１は極めて高い確率で訓練データテキスト１の「人間失格」に所属する、実際にも「人間失格」に所属する。

テストテキスト２は極めて高い確率で訓練データテキスト２の「単純ベイズ分類器の解説」に所属する、実際にも「単純ベイズ分類器の解説」に所属する。

テストテキスト１とテストテキスト２の分類結果は非常に良いと考えられる。

ただし、両方とも所属していないテストテキスト３は本モデルにより、約88.5%の確率で訓練データテキスト１の「人間失格」に所属する、約11.5%の確率で訓練データテキスト２の「単純ベイズ分類器の解説」に所属する。この結果は、実際の事実と異なっている。このテキストはインタネットで探している無関連の早口言葉であるので、用語として理論的な説明文より文学用語と近い可能性が高いと考えている。





---



**2.ナイーブベイズがどのようなときに有効で、どのようなときに有効ではないのか、考察しなさい** 

考察：

知能システム理論基礎の授業及び演習と自分の理解を通じて、私はナイーブベイズモデルは他の分類手法により誤差率が低いと思うが、ナイーブベイズの原理を考えると、ナイーブベイズモデルが各特徴がお互いに独立していることを前提と仮定している。ベイズの定理を用いて、各独立事相の確率により分類の確率を計算する。つまり、各特徴量が独立であるとき有効である。しかし、実際の分類問題中のデータには、各特徴が相関関係を持つが多いと思う、もし各特徴量お互いに関連性が高い特徴量が多い場合、ナイーブベイズモデルの有効性が低下となると考えている。





---



**3. 授業の感想を述べなさい**

　今回の演習授業で、Google Colaboratoryを利用して、実際のプログラミングすることで、具体的なテキストデータを用いてナイーブベイズモデルを利用して、分類タスクを体験した。実は理論基礎の授業で、渋谷先生挙げたgood badの例もナイーブベイズの原理をより詳しく理解した。今回の実験で、ナイーブベイズモデルの有効性について、もう一歩考えしながら、勉強になった。良かったと思っている。

　一方で、今回の授業で使っているGoogle Colaboratoryは、私昔の卒論研究でも深層学習モデルをモデリングするため、全部の実行や開発環境はColabを利用して開発していた。Colabは環境の設定が簡単し、GPU持たない学生・研究者も計算リソースを無料で利用することなど、特に現在TensorFlowだけでなく、Pytorchも簡単に利用できる、非常に便利な開発プラットフォームであると考えている。今後もこのサービスを利用して、研究や機械学習の勉強を続けたいと思っている。



---



**4. . [Advanced] 今回の演習で提供したライブラリでは、Scikit-learn でナイーブベイズを行った。Scikit-learn を使わずにナイーブベイズを実装しなさい。（必要であれば、形態素解析器として Janome を用いてもよい）**

まず、使用可能なライブラリをインストールする、今回はjanomeをインストールする。

In [1]:
!pip install janome
!pip install scipy

     |████████████████████████████████| 19.7 MB 1.4 MB/s 


その次、自分で準備したテキストデータを訓練用テキストデータとして使用するためアップロードする。

In [2]:
from google.colab import files
files.upload()

Saving ngsk.txt to ngsk.txt
Saving wiki.txt to wiki.txt


{'ngsk.txt': b'\xe7\xa7\x81\xe3\x81\xaf\xe3\x80\x81\xe3\x81\x9d\xe3\x81\xae\xe7\x94\xb7\xe3\x81\xae\xe5\x86\x99\xe7\x9c\x9f\xe3\x82\x92\xe4\xb8\x89\xe8\x91\x89\xe3\x80\x81\xe8\xa6\x8b\xe3\x81\x9f\xe3\x81\x93\xe3\x81\xa8\xe3\x81\x8c\xe3\x81\x82\xe3\x82\x8b\xe3\x80\x82\r\n\xe3\x80\x80\xe4\xb8\x80\xe8\x91\x89\xe3\x81\xaf\xe3\x80\x81\xe3\x81\x9d\xe3\x81\xae\xe7\x94\xb7\xe3\x81\xae\xe3\x80\x81\xe5\xb9\xbc\xe5\xb9\xb4\xe6\x99\x82\xe4\xbb\xa3\xe3\x80\x81\xe3\x81\xa8\xe3\x81\xa7\xe3\x82\x82\xe8\xa8\x80\xe3\x81\x86\xe3\x81\xb9\xe3\x81\x8d\xe3\x81\xa7\xe3\x81\x82\xe3\x82\x8d\xe3\x81\x86\xe3\x81\x8b\xe3\x80\x81\xe5\x8d\x81\xe6\xad\xb3\xe5\x89\x8d\xe5\xbe\x8c\xe3\x81\x8b\xe3\x81\xa8\xe6\x8e\xa8\xe5\xae\x9a\xe3\x81\x95\xe3\x82\x8c\xe3\x82\x8b\xe9\xa0\x83\xe3\x81\xae\xe5\x86\x99\xe7\x9c\x9f\xe3\x81\xa7\xe3\x81\x82\xe3\x81\xa3\xe3\x81\xa6\xe3\x80\x81\xe3\x81\x9d\xe3\x81\xae\xe5\xad\x90\xe4\xbe\x9b\xe3\x81\x8c\xe5\xa4\xa7\xe5\x8b\xa2\xe3\x81\xae\xe5\xa5\xb3\xe3\x81\xae\xe3\x81\xb2\xe3\x81\xa8\xe3\x81\

In [3]:
!ls

ngsk.txt  sample_data  wiki.txt


**課題の核心部分のコード：**

In [122]:
from janome.tokenizer import Tokenizer
import numpy as np

In [142]:
class HAONaiveBayes():
    #テキスト入力部
    def __init__(self, file1, file2):
        f = open(file1)
        self.text1 = f.read()
        f = open(file2)
        self.text2 = f.read()

        full_text = self.text1 + self.text1

        self.text1Dic = self.split_words(self.text1)
        self.text2Dic = self.split_words(self.text2)
    #テキスト分割部
    def split_words(self,text):
        t = Tokenizer()
        tokens = t.tokenize(text)
        List =  [token.surface for token in tokens]
        
        text = {}
        for i in List:
            if i not in text.keys():
                text[i] = 1
            else:
                text[i] += 1
        return text
    #ナイーブベイズ予測部
    def NaiveBayes(self,textTest):
        dic1 = self.text1Dic
        dic2 = self.text2Dic
        #辞書分割により初期化
        dicTest = self.split_words(textTest)
        #テキスト１とテキスト２に所属する確率の初期化
        P1 = 1/2
        P2 = 1/2

        for j in dicTest.keys():
            if j not in dic1.keys() or j not in dic2.keys():
                continue
        #ナイーブベイズにより確率を計算する
            P1 = P1 * np.log(dic1[j] / sum(dic1.values()))
            P2 = P2 * np.log(dic2[j] / sum(dic2.values()))

        #予測の出力
        P = P1/(P1+P2),P2/(P1+P2)
        return P
    
    def print_file1(self): #テキスト１の出力
        print(self.text1)
    def print_file2(self): #テキスト２の出力
        print(self.text2)


In [129]:
nb = HAONaiveBayes("ngsk.txt", "wiki.txt")

In [130]:
nb.print_file1()

私は、その男の写真を三葉、見たことがある。
　一葉は、その男の、幼年時代、とでも言うべきであろうか、十歳前後かと推定される頃の写真であって、その子供が大勢の女のひとに取りかこまれ、
（それは、その子供の姉たち、妹たち、それから、従姉妹いとこたちかと想像される）庭園の池のほとりに、荒い縞の袴はかまをはいて立ち、首を三十度ほど左に傾け、
醜く笑っている写真である。醜く？　けれども、鈍い人たち（つまり、美醜などに関心を持たぬ人たち）は、面白くも何とも無いような顔をして、
「可愛い坊ちゃんですね」
　といい加減なお世辞を言っても、まんざら空からお世辞に聞えないくらいの、謂いわば通俗の「可愛らしさ」みたいな影もその子供の笑顔に無いわけではないのだが、
しかし、いささかでも、美醜に就いての訓練を経て来たひとなら、ひとめ見てすぐ、
「なんて、いやな子供だ」
　と頗すこぶる不快そうに呟つぶやき、毛虫でも払いのける時のような手つきで、その写真をほうり投げるかも知れない。
　まったく、その子供の笑顔は、よく見れば見るほど、何とも知れず、イヤな薄気味悪いものが感ぜられて来る。どだい、それは、笑顔でない。
この子は、少しも笑ってはいないのだ。その証拠には、この子は、両方のこぶしを固く握って立っている。人間は、こぶしを固く握りながら笑えるものでは無いのである。
猿だ。猿の笑顔だ。ただ、顔に醜い皺しわを寄せているだけなのである。「皺くちゃ坊ちゃん」とでも言いたくなるくらいの、まことに奇妙な、そうして、
どこかけがらわしく、へんにひとをムカムカさせる表情の写真であった。私はこれまで、こんな不思議な表情の子供を見た事が、いちども無かった。



In [131]:
nb.print_file2()

概要
　単純ベイズ分類器の元となる確率モデルは強い（単純な）独立性仮定と共にベイズの定理を適用することに基づいており、
より正確に言えば「独立特徴モデル」と呼ぶべきものである。

　確率モデルの性質に基づいて、単純ベイズ分類器は教師あり学習の設定で効率的に訓練可能である。多くの実用例では、
単純ベイズ分類器のパラメータ推定には最尤法が使われる。つまり、単純ベイズ分類器を使用するにあたって、ベイズ確率やその他のベイズ的手法を使う必要はない。

　設計も仮定も非常に単純であるにも関わらず、単純ベイズ分類器は複雑な実世界の状況において、期待よりもずっとうまく働く。
近頃、ベイズ分類問題の注意深い解析によって、単純ベイズ分類器の効率性に理論的理由があることが示された。
単純ベイズ分類器の利点は、分類に不可欠なパラメータ（変数群の平均と分散）を見積もるのに、訓練例データが少なくて済む点である。
変数群は独立であると仮定されているため、各クラスについての変数の分散だけが必要であり、共分散行列全体は不要である。



In [143]:
test1 = "第二葉の写真の顔は、これはまた、びっくりするくらいひどく変貌していた。"
print(nb.NaiveBayes(test1))

(0.3488865767129076, 0.6511134232870923)


In [144]:
test2 = "抽象的には、分類器の確率モデルは次のような依存クラス変数Cについての条件付きモデルである。"
print(nb.NaiveBayes(test2))

(0.5245994339101785, 0.47540056608982156)


In [145]:
test3 = "斜め七十七度の並びで泣く泣くいななくナナハン七台難なく並べて長眺め"
print(nb.NaiveBayes(test3))

(0.4805228803689696, 0.5194771196310303)


分類の結果として、

　テストテキスト1は約34.9%の確率で訓練データテキスト1の「人間失格」に所属する、65.1%確率で訓練データテキスト2の「単純ベイズ分類器の解説」に所属すると予測したが、実際には訓練データテキスト1「人間失格」に所属する。

　テストテキスト2は約52.5%の確率で訓練データテキスト2の「人間失格」に所属する、47.5%確率で訓練データテキスト2の「単純ベイズ分類器の解説」に所属すると予測したが、実際には訓練データテキスト2の「単純ベイズ分類器の解説」に所属する。

　両方とも所属していないテストテキスト3は本モデルにより、約48.1%の確率で訓練データテキスト1の「人間失格」に所属する、約51.9%の確率で訓練データテキスト3の「単純ベイズ分類器の解説」に所属する。

　この結果を通じて、sklearnライブラリを使わずに自分で作ったナイーブベイズライブラリを用いて分類器の分類効果はsklearnライブラリを使う分類器により、予想以外低いと考えている。レポート課題の時間切まで改良できなかったが、今後分類精度を上げる方法などを検討したいと考える。



---

